### PR TITLE
8308881: Strong CLD oop handle roots are demoted to non-roots concurrently

### PR DESCRIPTION
**Pull request by fisk**
It is illegal to remove strong roots concurrently without clearing them first. A SATB collector with concurrent root scanning assumes that when strong roots disappear from the object graph, they are cleared first, which makes SATB notice the root. All global strong roots do this. Except CLD strong roots, which are turned into non-roots by decrementing the keep_alive counter to 0, when bootstrapping weak hidden class CLDs. This is not valid behaviour. This patch tries to treat these oops like we do any other global strong handles that are unlinked from the system: clear them when they stop being strong roots.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8308881](https://bugs.openjdk.org/browse/JDK-8308881): Strong CLD oop handle roots are demoted to non-roots concurrently

### Reviewers
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [b75fce32](https://git.openjdk.org/jdk/pull/14154/files/b75fce3261c5c49d3f1dee07daa66d8fd3766c5f)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14154/head:pull/14154` \
`$ git checkout pull/14154`

Update a local copy of the PR: \
`$ git checkout pull/14154` \
`$ git pull https://git.openjdk.org/jdk.git pull/14154/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14154`

View PR using the GUI difftool: \
`$ git pr show -t 14154`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14154.diff">https://git.openjdk.org/jdk/pull/14154.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14154#issuecomment-1563135343)

### CODE REVIEW

**Review by stefank — COMMENTED** [review]
Could this be using CollectedHeap::keep_alive instead of performing this clear/store dance?

**Comment by fisk** [comment]
> Could this be using CollectedHeap::keep_alive instead of performing this clear/store dance?

No, because the problem isn't keeping the object alive. It's already kept alive using handles. The problem is that the snapshot of strong roots are not processed during marking, leading to ABA problems, causing things to go wrong later on. Therefore I wanted to model the solution conceptually as close as possible to other handles. Admittedly, it does look more like an exptic rain dance than ideal. Just got to clap our hands, spin around 360 degrees and then jump - then clap our hands again!

**Review by coleenp — APPROVED** [review]
This is strange but if it works, it's ok with me.

**Comment by fisk** [comment]
Thanks for the review @coleenp!

File: src/hotspot/share/classfile/classLoaderData.cpp [hunk 1]
```diff
@@ -300,6 +300,30 @@ bool ClassLoaderData::try_claim(int claim) {
...
+  // The way we would retire a strong root, is by clearing it. Then its address can be
+  // reused for a new handle, that isn't a normal strong root. We do that dance below.
+  class TransitionRootsOopClosure : public OopClosure {
+  public:
+    virtual void do_oop(oop* p) {
+      oop obj = NativeAccess<>::oop_load(p); // Load the strong root
+      NativeAccess<>::oop_store(p, nullptr); // Clear the strong root
+      // No-op free handle
+      // No-op allocate new handle using the same address
+      NativeAccess<>::oop_store(p, obj); // Store the strong non-root
```
Comment (xmas92) [inline src/hotspot/share/classfile/classLoaderData.cpp:317]: Is there an issue if someone with an `OopHandle(p)` racingly reads null while a root is transitioning?

Comment (fisk) [reply src/hotspot/share/classfile/classLoaderData.cpp:317]: Shouldn't be, but I'll remove the null store and write some more comments, to be on the safe side.

**Review by stefank — APPROVED** [review]

**Comment by xmas92** [comment]
Is there any issue with memory order access here?
From my understanding this works because a GC which starts before `demote_strong_roots` has the property if it observers `_keep_alive == 0` then in must have observed `demote_strong_roots` (either root throughout or it observed the transition through barriers). And if a GC starts after `demote_strong_roots` it must never observe `_keep_alive != 0`.

**Comment by fisk** [comment]
> Is there any issue with memory order access here?
> From my understanding this works because a GC which starts before `demote_strong_roots` has the property if it observers `_keep_alive == 0` then in must have observed `demote_strong_roots` (either root throughout or it observed the transition through barriers). And if a GC starts after `demote_strong_roots` it must never observe `_keep_alive != 0`.

If the GC races with these accesses, it doesn't matter if the race hits before or after the accesses. The accesses themselves will sync with the GC. To be precise:
1. If the object relocated, the load barrier will detect that and the new value stored will pointed to the to-space object.
2. The store barrier will load the previous value and enqueue for marking, so that we don't miss there was an oop in there that needed to be marked.
3. After the last store, the colors will be right with both ZGC collectors. GC will never perform non-monotonic transitions away from the best color, which the store ensures we will get.

If the GC processes the location before these accesses, it will ensure the above properties. Whoever gets there first doesn't matter as long as one of them will.

**Comment by fisk** [comment]
And whether the GC observes the old or new value of the keep alive counter doesn't matter either, as at least one thread ensures the oop location gets processed. If the GC sees the counter as 0 in this race, the mutator will process the oop location as I described. And then at least one thread has processed it.

**Pull request closed without merging**